accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,449 @@
+
+```dart
+extension MyFancyList<T> on List<T> {
+  int get doubleLength => this.length * 2;
+  List<T> operator-() => this.reversed.toList();
+  List<List<T>> split(int at) => 
+      <List<T>>[this.sublist(0, at), this.sublist(at)];
+  List<T> mapToList<R>(R Function(T) convert) => this.map(convert).toList();
+}
+```
+
+More precisely, an extension declaration is a top-level declaration with a grammar similar to:
+
+```ebnf
+<extension> ::= 
+  `extension' <identifier> <typeParameters>? `on' <type> `?'? `{'
+     memberDeclaration*
+  `}'
+```
+
+Such a declaration introduces its *name* (the identifier) into the surrounding scope. The name does not denote a type, but it can be used to denote the extension itself in various places. The name can be hidden or shown in `import` or `export` declarations.
+
+The *type* can be any valid Dart type, including a single type variable. It can refer to the type parameters of the extension. It can be followed by `?` which means that it allows `null` values. When Dart gets non-nullable types by default (NNBD), this `?` syntax is removed and subsumed by nullable types like `int?` being allowed in the `<type>` position.
+
+The member declarations can be any static or instance member declaration except for instance variables and constructors.
+
+### Scope
+
+Dart's static extension methods are *scoped*. They only apply to code that the extension itself is accessible to. Being accessible means that using the extension's name must denote the extension. The extension is not in scope if another declaration with the same name shadows the extension, if the extension's library is not imported, if the library is imported and the extension is hidden, or the library is only imported with a prefix. In other words, if the extension had been a class, it is only in scope if using the name would denote the class&mdash; which will allow you to call static methods on the class, which is exactly what we are going to do.
+
+### Extension Member Resolution
+
+The declaration introduces an extension. The extension's `on` type defines which types are being extended.
+
+For any member access, `x.foo`, `x.bar()`, `x.baz = 42`, `x(42)` or `x + y`, the language first checks whether the static type of `x` has a member with the same base name as the operation. That is, if it has a corresponding instance member, respectively, a `foo` method or getter or a `foo=` setter. a `bar` member or `bar=` setter, a `baz` member or `baz=` setter, a `call` method, or a `+` method. If so, then the operation is unaffected by extensions. *This check does not care whether the invocation is otherwise correct, based on number or type of the arguments, it only checks if there is a member at all.*
+
+(The types `dynamic` and `Never` are considered as having all members, the type `void` is always a compile-time error when used in a receiver position, so none of these can ever be affected by static extension methods).
```

Comment:
In discussions with @srawlins today, some use cases came up that made me question whether treating dynamic calls as never resolving to an extension is really the right thing to do.  The concrete example was JSON, where you could imagine defining a package which adds `.toInt` etc to `dynamic`.  It also occurred to me that if you use extension methods to define things like kotlin's `let`, `apply`, and `also`, it's a bit weird that you can't use them on things of type dynamic.  Is it really useful to keep extensions that apply to everything from applying to dynamic?

Reply:
The design is based on the idea that existing methods takes precedence over extension methods because that way adding extension methods cannot break existing code. 
It will allow us to add extension methods to the platform libraries, and not break anyone who uses an existing member with the same name on their own subtype of the type.

That reasoning should apply to dynamic invocations too. If someone writes `dynamicValue.foo` anywhere, we should not let an `extension o on Object? { int get foo => 42; }` override their existing behavior.

If we are willing to drop that protection, maybe we can do so everywhere, and let extension methods take precedence over existing methods. That would be a very different feature (one which allows overriding `String` methods).
It will probably also mean that we cannot add extension methods on any subclassable platform type to any existing platform library.